process/block/preprocess/transactions.go

```diff
@@ -472,151 +718,151 @@ func (txs *transactions) getAllTxsFromMiniBlock(
-	orderedTxs, orderedTxHashes, err := txs.computeOrderedTxs(senderShardId, receiverShardId)
-	timeAfter := time.Now()
+	mapMiniBlocks := make(map[uint32]*block.MiniBlock)
 
-	if err != nil {
-		log.Trace("computeOrderedTxs", "error", err.Error())
-		return nil, err
-	}
+	numTxsAdded := 0
+	numTxsBad := 0
+	numTxsSkipped := 0
 
-	if !haveTime() {
-		log.Debug("time is up ordering txs",
-			"num txs", len(orderedTxs),
-			"time [s]", timeAfter.Sub(timeBefore).Seconds(),
-		)
-		return nil, process.ErrTimeIsOut
-	}
+	totalTimeUsedForProcesss := time.Duration(0)
+	totalTimeUsedForComputeGasConsumed := time.Duration(0)
 
-	log.Trace("time elapsed to ordered txs",
-		"num txs", len(orderedTxs),
-		"time [s]", timeAfter.Sub(timeBefore).Seconds(),
-	)
+	firstInvalidTxFound := false
 
-	miniBlock := &block.MiniBlock{}
-	miniBlock.SenderShardID = senderShardId
-	miniBlock.ReceiverShardID = receiverShardId
-	miniBlock.TxHashes = make([][]byte, 0)
-	miniBlock.Type = block.TxBlock
+	gasConsumedByMiniBlocksInSenderShard := uint64(0)
+	mapGasConsumedByMiniBlockInReceiverShard := make(map[uint32]uint64)
+	totalGasConsumedInSelfShard := txs.gasHandler.TotalGasConsumed()
 
-	addedTxs := 0
-	gasConsumedByMiniBlockInSenderShard := uint64(0)
-	gasConsumedByMiniBlockInReceiverShard := uint64(0)
+	senderAddressToSkip := []byte("")
 
-	for index := range orderedTxs {
-		txHandler := orderedTxs[index]
-		tx := txHandler.(*transaction.Transaction)
-		txHash := orderedTxHashes[index]
+	defer func() {
+		go txs.notifyTransactionProviderIfNeeded()
+	}()
 
+	for index := range sortedTxs {
 		if !haveTime() {
+			log.Debug("time is out in createAndProcessMiniBlock")
+			break
+		}
+
+		txHandler := sortedTxs[index].Tx
+		tx := txHandler.(*transaction.Transaction)
+		txHash := sortedTxs[index].TxHash
+		senderShardID := sortedTxs[index].SenderShardID
+		receiverShardID := sortedTxs[index].ReceiverShardID
+
+		numNewMiniBlocks := 0
+		if _, ok := mapMiniBlocks[receiverShardID]; !ok {
+			numNewMiniBlocks = 1
+		}
+		if txs.blockSizeComputation.IsMaxBlockSizeReached(numNewMiniBlocks, 1) {
+			log.Debug("max txs accepted in one block is reached",
+				"num txs added", numTxsAdded,
+				"total txs", len(sortedTxs))
 			break
 		}
 
-		if txs.isTxAlreadyProcessed(txHash, &txs.txsForCurrBlock) {
+		if isShardStuck != nil && isShardStuck(receiverShardID) {
+			log.Trace("shard is stuck", "shard", receiverShardID)
 			continue
 		}
 
+		if len(senderAddressToSkip) > 0 {
+			if bytes.Equal(senderAddressToSkip, tx.GetSndAddress()) {
+				numTxsSkipped++
+				continue
+			}
+		}
+
 		snapshot := txs.accounts.JournalLen()
-		oldGasConsumedByMiniBlockInSenderShard := gasConsumedByMiniBlockInSenderShard
-		oldGasConsumedByMiniBlockInReceiverShard := gasConsumedByMiniBlockInReceiverShard
 
-		err = txs.computeGasConsumed(
-			miniBlock.SenderShardID,
-			miniBlock.ReceiverShardID,
+		gasConsumedByMiniBlockInReceiverShard := mapGasConsumedByMiniBlockInReceiverShard[receiverShardID]
```

Comment:
this function is way too big, could we separate it into some smaller pieces ?

Reply:
Analyzed but it would result in methods with 6-7 pointer parameters as from the whole for it is called many times continue or break, which reverts the state of many variables